crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -890,45 +890,67 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
         }
         let dir_entry = table.get_dir(dirfd)?;
 
+        let oflags = OFlags::from(&oflags);
+
+        // OFlags that are not supported for directories
+        let non_dir_oflags = oflags.contains(OFlags::CREATE)
+            || oflags.contains(OFlags::EXCLUSIVE)
+            || oflags.contains(OFlags::TRUNCATE);
+
+        if oflags.contains(OFlags::DIRECTORY) && non_dir_oflags {
+            return Err(Error::invalid_argument().context("incompatible oflag combination"));
+        }
+
+        let mut required_caps = DirCaps::OPEN;
+        if oflags.contains(OFlags::CREATE) {
+            required_caps = required_caps | DirCaps::CREATE_FILE;
+        }
+        let dir = dir_entry.get_cap(required_caps)?;
+
         let symlink_follow = dirflags.contains(types::Lookupflags::SYMLINK_FOLLOW);
 
-        let oflags = OFlags::from(&oflags);
-        let fdflags = FdFlags::from(fdflags);
         let path = path.as_str()?;
-        if oflags.contains(OFlags::DIRECTORY) {
-            if oflags.contains(OFlags::CREATE)
-                || oflags.contains(OFlags::EXCLUSIVE)
-                || oflags.contains(OFlags::TRUNCATE)
-            {
-                return Err(Error::invalid_argument().context("directory oflags"));
+        let path = path.deref();
+        match dir.get_path_filestat(path, symlink_follow).await {
```

Comment:
I am concerned about this `get_path_filestat`, because it does a second path lookup. There's a window for a race condition here, where a directory could be replaced by something which is not a directory. I don't yet have a clear view of the consequences of the open succeeding on an unexpected type, so this is something I want to investigate more.

I wonder if it would be possible to instead do something like:
 - Attempt to open with open_dir; if that works, create a `DirEntry` entry.
 - If that fails with `Notdir`, open it with regular `open`.
Would that work?

I know that this is redundant on POSIX hosts, and possibly we could optimize things even more there, but this code is currently written to support Windows also.

Reply:
I now have an implementation of this here: https://github.com/bytecodealliance/wasmtime/pull/4967

It uses the same testcase as this PR, so it fixes the same bug, but does so in a way that doesn't require doing a stat followed by a separate open. There's more room for improvement beyond that, but hopefully that's a good step.